lib/plugins/aws/lib/updateStack.js

```diff
@@ -33,43 +34,72 @@ module.exports = {
       stackTags = Object.assign(stackTags, this.serverless.service.provider.stackTags);
     }
 
-    const params = {
+    const createChangeSetParams = {
       StackName: stackName,
-      OnFailure: 'DELETE',
+      ChangeSetName: changeSetName,
+      ChangeSetType: 'CREATE',
       Capabilities: ['CAPABILITY_IAM', 'CAPABILITY_NAMED_IAM'],
       Parameters: [],
       TemplateURL: templateUrl,
       Tags: Object.keys(stackTags).map((key) => ({ Key: key, Value: stackTags[key] })),
     };
 
+    const executeChangeSetParams = {
+      StackName: stackName,
+      ChangeSetName: changeSetName,
+    };
+
     if (
       this.serverless.service.provider.compiledCloudFormationTemplate &&
       this.serverless.service.provider.compiledCloudFormationTemplate.Transform
     ) {
-      params.Capabilities.push('CAPABILITY_AUTO_EXPAND');
+      createChangeSetParams.Capabilities.push('CAPABILITY_AUTO_EXPAND');
     }
 
     const customDeploymentRole = this.provider.getCustomDeploymentRole();
     if (customDeploymentRole) {
-      params.RoleARN = customDeploymentRole;
+      createChangeSetParams.RoleARN = customDeploymentRole;
     }
 
     if (this.serverless.service.provider.notificationArns) {
-      params.NotificationARNs = this.serverless.service.provider.notificationArns;
+      createChangeSetParams.NotificationARNs = this.serverless.service.provider.notificationArns;
     }
 
     if (this.serverless.service.provider.stackParameters) {
-      params.Parameters = this.serverless.service.provider.stackParameters;
+      createChangeSetParams.Parameters = this.serverless.service.provider.stackParameters;
     }
 
     if (this.serverless.service.provider.disableRollback) {
-      delete params.OnFailure;
-      params.DisableRollback = this.serverless.service.provider.disableRollback;
+      executeChangeSetParams.DisableRollback = this.serverless.service.provider.disableRollback;
     }
 
-    return this.provider
-      .request('CloudFormation', 'createStack', params)
-      .then((cfData) => this.monitorStack('create', cfData));
+    // Create new change set
+    log.info('Creating new change set');
+    const changeSet = await this.provider.request(
+      'CloudFormation',
+      'createChangeSet',
+      createChangeSetParams
+    );
+
+    // Wait for changeset to be created
+    log.info('Waiting for new change set to be created');
+    const changeSetDescription = await this.waitForChangeSetCreation(changeSetName, stackName);
+
+    // Check if stack has changes
+    if (isChangeSetWithoutChanges(changeSetDescription)) {
+      // Cleanup changeset when it does not include any changes
+      log.info('Created change set does not include any changes, removing it');
+      await this.provider.request('CloudFormation', 'deleteChangeSet', {
```

Comment:
I wonder do we need to remove changeset in this case? At least if I see correctly we're not removing it after deployment if there are changes to apply. Why we treat those two cases differently for that matter?

Reply:
When change set is executed, it it automatically removed so we don't have to clean it up manually

Reply:
Thanks, that clarifies 👍